---
params:
  tbprofiler: "all.txt"
  snp_diff_matrix: "all_samples_snp_diff.txt"
  snp_diff_heatmap: "all_samples_snp_diff_heatmap.png"
  phylotree: "Phylogenetic_Tree.png"
  kraken: "kraken.result"
  percentage_MTBC_threshold: "90"
  mixinfect: "ALL_MixedSampleSummary.txt"
title: "<center>TBSeqPipe Analysis Summary Report</center>"
date: "<center>Date of Report：`r format(Sys.time(), '%Y/%m/%d')`</center>"
output:
  html_document:
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: no
    toc_depth: 3
    theme: cerulean
    highlight: tango
    body_placement: left
  pdf_document:
    toc: yes
    toc_depth: '3'
---

```{r load packages, echo = FALSE, message = FALSE, warning = FALSE}
library(DT)
```

***

# 1. Introduction

Tuberculosis (TB) is an infectious disease usually caused by *Mycobacterium tuberculosis* complex(MTBC) bacteria, and still is a threat to global health. Rapid, reliable and increasingly affordable WGS technologies can now guide all components of TB control: diagnosis, treatment, surveillance and source investigation.

Analysis of this workflow starts from illumina WGS data of Mycobacterium Tuberculosis complex isolates. Several basic and downstream high-level analysis tasks are performed. Main analysis modules include reads preprocessing, MTBC identification, reads mapping and variants calling, lineage and drug resistance profiling, SNP-distance calculation, phylogenetic analysis, and mixed infection detection. Methods and results of these modules are integrated and presented here in the summary report.

&nbsp;

# 2. Methods and Results

## 2.1 Reads Preprocessing

Fastp is used for trimming low-quality bases from 5' and/or 3' ends of each read and removing adapter sequences. Reads quality control is performed using FastQC, which generates a quality report for each FASTQ file, containing information such as the per base and average sequence quality (using the Phred quality score), overrepresented sequences, GC content, adapter and the k-mer content of the FASTQ file. Finally, MultiQC is used to aggregate the FastQC reports for a given set of FASTQ files into a single report, allowing reviews of all FASTQ files at once.

[Reads QC report](../00.data_cleaning/multiqc/multiqc_report.html)

&nbsp;

**Result files**: [Reads Preprocessing](../00.data_cleaning) <br>

&nbsp;

## 2.2 MTBC Identification

To filter out samples that may have been contaminated by foreign DNA during sample preparation, the reads for each sample are inspected through Kraken software against the MiniKraken database containing complete bacterial, archaeal, and viral genomes and the GRCh38 human genome from the Refseq database. Any sample will be dropped if less than `r params$percentage_MTBC_threshold`% of its reads were taxonomically classified under the *Mycobacterium tuberculosis* complex (MTBC).

<center>**Table 1. MTBC Identification Results**</center>

```{r echo=FALSE, message=FALSE}
kraken <- read.table(
  params$kraken,
  header = F,
  na.strings = c("NA"),
  sep = "\t",
  comment.char = "#"
)

colnames(kraken) <- c("sample_id", "percentage_MTBC (%)", "category_detected")

datatable(
  kraken,
  rownames = FALSE,
  extensions = "FixedColumns",
  option = list(
    scrollX = TRUE,
    scrollY = "300px",
    paging = FALSE
  )
) %>%
  formatStyle(
    'percentage_MTBC (%)',
    color = styleInterval(c(params$percentage_MTBC_threshold), c('red', 'black'))
  )
```

<font size="1">**Note**: percentage_MTBC: the percentage of the reads that are taxonomically classified under the *Mycobacterium tuberculosis* Complex (MTBC); category_detected：the classification result with the highest proportion of reads.</font>

&nbsp;

**Result files**: [MTBC Identification](../00.data_cleaning) <br>

&nbsp;

## 2.3 Reads Mapping and Variants Calling

### 2.3.1 Reads Mapping

Sequencing reads after preprocessing are mapped to the reference genome (H37Rv,NC000962.3) with BWA MEM. Duplicate reads are marked and deduplicated using SAMtools, and then the Genome Analysis Toolkit (GATK) is used to realign and recalibrate base quality.

&nbsp;

### 2.3.2 Alignment Quality Control

The results of the quality control of all alignment bam files generated by Qualimap are assembled using the MultiQC.

[Alignment QC report](../01.alignment/multiqc/multiqc_report.html)

&nbsp;

**Result files**: [Reads Mapping and Variants Calling - Reads Mapping](../01.alignment) <br>

&nbsp;

### 2.3.3 Variants Calling and Filtering

SAMtools/BCFtools suite is used for calling and filtering SNPs and short INDELs. Two different types of SNPs are detected: fSNP (fixed SNP for phylogenetic analysis) and vSNP (SNPs with low frequency are also included for Drug resistance and mixed infection detection). SNPs in repetitive regions of the genome (e.g.PPE/PE-PGRS family genes, phagesequence,insertion or mobile genetic elements) are filtered out.

&nbsp;

**Result files**: [Reads Mapping and Variants Calling - Variants Calling](../02.snp_calling) <br>

&nbsp;

## 2.4 Lineage and Drug Resistance Profiling

TB-Profiler is the tool in this workflow for lineage and drug-resistance profiler based on reliable mutation libraries.

<center>**Table 2. Lineage and Drug Resistance Results**</center>

```{r echo=FALSE, message=FALSE}
tbprofiler <- read.table(
  params$tbprofiler,
  header = T,
  na.strings = c("NA"),
  sep = "\t"
)

colnames(tbprofiler)[1] <- "sample_id"

datatable(
  tbprofiler,
  rownames = FALSE,
  extensions = "FixedColumns",
  option = list(
    scrollX = TRUE,
    scrollY = "300px",
    paging = FALSE,
    fixedColumns = list(leftColumns = 1)
  )
)
```

<font size="1">**Note**: DR_type：drug-resistance classification result; num_dr_variants：number of drug-resistance associated variants; num_other_variants：number of variants in drug-resistance associated genes but not proved to be associated with drug-resistance.</font>

&nbsp;

**Result files**: [Lineage and Drug Resistance Profiling](../05.lineage_dr) <br>

&nbsp;

## 2.5 SNP-distance Calculation

An in-house script is used to calculate relative pairwise SNP distance among all samples. Additionally, in this analysis, SNPs in drug-resistance associated genes are removed to better represent genetic distance.

### 2.5.1 Pairwise SNP-distance Matrix

A matrix is generated to present all pairwise SNP-distance among all samples.

<center>**Table 3. Pairwise SNP-distance matrix**</center>

```{r echo=FALSE, message=FALSE}
snp_dist <- read.table(
  params$snp_diff_matrix,
  header = T,
  na.strings = c("NA"),
  sep = "\t",
  check.names = F
)

colnames(snp_dist)[1] <- "sample_id"

datatable(
  snp_dist,
  rownames = FALSE,
  extensions = "FixedColumns",
  option = list(
    scrollX=TRUE,
    scrollY="300px",
    paging=FALSE,
	fixedColumns = list(leftColumns = 1)
  )
)
```

&nbsp;

### 2.5.2 Pairwise SNP-distance Heatmap

A heatmap is used for the pairwise SNP-distance visualization. Lineage and drug-resistance results are also added into the heatmap.

```{r echo=FALSE, out.width="80%", fig.align="center"}
knitr::include_graphics(params$snp_diff_heatmap)
```

<center>**Figure 1. Pairwise SNP-distance Heatmap**</center>

&nbsp;

**Result files**: [SNP-distance Calculation](../04.snp_distance) <br>

&nbsp;

## 2.6 Phylogenetic Analysis

The SNPs for SNP-distance calculation are combined into a single consensus and non-redundant list. The alignments of the polymorphic positions in the strains with *M. canettii* (CIPT 140060008) added as root are used for maximum likelihood tree reconstruction using RAxML-NG with '-model GTR+G'. R-ggtree is then used for phylogenetic tree visualization with the
results of lineage and drug-resistance profiling.

```{r echo=FALSE, out.width="100%", fig.align="center", dpi=1000}
knitr::include_graphics(params$phylotree)
```

<center>**Figure 2. Phylogenetic Tree Visulization**</center>

&nbsp;

**Result files**: [Phylogenetic Analysis](../06.phylotree) <br>

&nbsp;

## 2.7 Mixed Infection Detection

MixInfect is used to identify if mixed infection exists in each sample based on WGS data. MixInfect applies the method of Bayesian model-based clustering of allele frequencies from sequencing reads at heterozygous sites to detect mixed infection.

<center>**Table 4. Mixed Infection Detection Results**</center>

```{r echo=FALSE, message=FALSE}
mixinfect <- read.table(
  params$mixinfect,
  header = F,
  na.strings = c("NA"),
  sep = "\t",
  comment.char = "#"
)

mixinfect <- mixinfect[c(1,2,6,7)]

colnames(mixinfect) <- c("sample_id", "mix", "n_mixed_strains", "major_strain_proportion")

datatable(
  mixinfect,
  rownames = FALSE,
  extensions = "FixedColumns",
  option = list(
    scrollX = TRUE,
    scrollY = "300px",
    paging = FALSE
  )
)
```

<font size="1">**Note**: mix：mix infection detection result（Mix：detected; Non-mix：not detected）; n_mixed_strains：if mixed infection is detected, number of different strains in the mixed sample (3 in maximum); major_strain_proportion：if mixed infection is detected, proportion of the major strain in the mixed sample.</font>

&nbsp;

**Result files**: [Mixed Infection Detection](../07.mix_infection) <br>

&nbsp;

# 3. References

[1] Meehan, Conor J et al. “Whole genome sequencing of Mycobacterium tuberculosis: current standards and open issues.” Nature reviews. Microbiology vol. 17,9 (2019): 533-545. doi:10.1038/s41579-019-0214-5

[2] Yang, Tingting et al. “SAM-TB: a whole genome sequencing data analysis website for detection of Mycobacterium tuberculosis drug resistance and transmission.” Briefings in bioinformatics vol. 23,2 (2022): bbac030. doi:10.1093/bib/bbac030

[3] Andrews S. (2010). FastQC: a quality control tool for high throughput sequence data. Available online at: http://www.bioinformatics.babraham.ac.uk/projects/fastqc

[4] Okonechnikov, Konstantin et al. “Qualimap 2: advanced multi-sample quality control for high-throughput sequencing data.” Bioinformatics (Oxford, England) vol. 32,2 (2016): 292-4. doi:10.1093/bioinformatics/btv566

[5] Ewels, Philip et al. “MultiQC: summarize analysis results for multiple tools and samples in a single report.” Bioinformatics (Oxford, England) vol. 32,19 (2016): 3047-8. doi:10.1093/bioinformatics/btw354

[6] Wood, Derrick E, and Steven L Salzberg. “Kraken: ultrafast metagenomic sequence classification using exact alignments.” Genome biology vol. 15,3 R46. 3 Mar. 2014, doi:10.1186/gb-2014-15-3-r46

[7] Chen, Shifu et al. “fastp: an ultra-fast all-in-one FASTQ preprocessor.” Bioinformatics (Oxford, England) vol. 34,17 (2018): i884-i890. doi:10.1093/bioinformatics/bty560

[8] Li, Heng, and Richard Durbin. “Fast and accurate short read alignment with Burrows-Wheeler transform.” Bioinformatics (Oxford, England) vol. 25,14 (2009): 1754-60. doi:10.1093/bioinformatics/btp324

[9] McKenna, Aaron et al. “The Genome Analysis Toolkit: a MapReduce framework for analyzing next-generation DNA sequencing data.” Genome research vol. 20,9 (2010): 1297-303. doi:10.1101/gr.107524.110

[10] Danecek, Petr et al. “Twelve years of SAMtools and BCFtools.” GigaScience vol. 10,2 (2021): giab008. doi:10.1093/gigascience/giab008

[11] Cingolani, Pablo et al. “A program for annotating and predicting the effects of single nucleotide polymorphisms, SnpEff: SNPs in the genome of Drosophila melanogaster strain w1118; iso-2; iso-3.” Fly vol. 6,2 (2012): 80-92. doi:10.4161/fly.19695

[12] Phelan, Jody E et al. “Integrating informatics tools and portable sequencing technology for rapid detection of resistance to anti-tuberculous drugs.” Genome medicine vol. 11,1 41. 24 Jun. 2019, doi:10.1186/s13073-019-0650-x

[13] Coll, Francesc et al. “Genome-wide analysis of multi- and extensively drug-resistant Mycobacterium tuberculosis.” Nature genetics vol. 50,2 (2018): 307-316. doi:10.1038/s41588-017-0029-0

[14] Kozlov, Alexey M et al. “RAxML-NG: a fast, scalable and user-friendly tool for maximum likelihood phylogenetic inference.” Bioinformatics (Oxford, England) vol. 35,21 (2019): 4453-4455. doi:10.1093/bioinformatics/btz305

[15] Yu, Guangchuang et al. “Two Methods for Mapping and Visualizing Associated Data on Phylogeny Using Ggtree.” Molecular biology and evolution vol. 35,12 (2018): 3041-3043. doi:10.1093/molbev/msy194

[16] Sobkowiak, Benjamin et al. “Identifying mixed Mycobacterium tuberculosis infections from whole genome sequence data.” BMC genomics vol. 19,1 613. 14 Aug. 2018, doi:10.1186/s12864-018-4988-z
